Reference leaks in test_expression

When running with a debug build of Python (specifically, this was Python 3.9b0), I'm seeing object reference leaks from `test_expression`.

```
test_expression leaked [44, 44, 44, 44] references, sum=176
test_expression leaked [12, 12, 12, 12] memory blocks, sum=48
0:00:23 load avg: 2.36 [28/90/2] test_extended_notifiers -- test_expression failed
```

Note: since this was with a Python pre-release, the first thing we should do it retest with a released version of Python, just in case: this could conceivably be a core Python bug (unlikely, but possible).

See #725 for steps to run a reference leak test.

Reproduced this with Python 3.8.3 built from source. So it's unlikely to be a core Python bug.

I also see some very strange failures from the refleak test run, presumably resulting from running the relevant tests multiple times:

```
======================================================================
FAIL: test_dict_event_repr (test_trait_dict_list_set_event.TestTraitEvent)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/mdickinson/Enthought/ETS/traits/traits/tests/test_trait_dict_list_set_event.py", line 53, in test_dict_event_repr
    self.assertEqual(event.__repr__(), event_str)
AssertionError: "TraitDictEvent(removed={}, added={}, changed={'blue': 10, 'black': 0})" != "TraitDictEvent(removed={}, added={'black': 0}, changed={'blue': 0})"
- TraitDictEvent(removed={}, added={}, changed={'blue': 10, 'black': 0})
?                                                       -------------
+ TraitDictEvent(removed={}, added={'black': 0}, changed={'blue': 0})
?                                   ++++++++++


======================================================================
FAIL: test_list_repr (test_trait_dict_list_set_event.TestTraitEvent)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/mdickinson/Enthought/ETS/traits/traits/tests/test_trait_dict_list_set_event.py", line 41, in test_list_repr
    self.assertEqual(event.__repr__(), event_str)
AssertionError: 'TraitListEvent(index=slice(0, 3, 2), removed=[4, 5], added=[4, 5])' != 'TraitListEvent(index=slice(0, 3, 2), removed=[1, 3], added=[4, 5])'
- TraitListEvent(index=slice(0, 3, 2), removed=[4, 5], added=[4, 5])
?                                               ^  ^
+ TraitListEvent(index=slice(0, 3, 2), removed=[1, 3], added=[4, 5])
?                                               ^  ^


======================================================================
FAIL: test_set_event_repr (test_trait_dict_list_set_event.TestTraitEvent)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/mdickinson/Enthought/ETS/traits/traits/tests/test_trait_dict_list_set_event.py", line 64, in test_set_event_repr
    self.assertEqual(event.__repr__(), event_str)
AssertionError: 'TraitSetEvent(removed={4}, added={3})' != 'TraitSetEvent(removed={3}, added={4})'
- TraitSetEvent(removed={4}, added={3})
?                        ^          ^
+ TraitSetEvent(removed={3}, added={4})
?                        ^          ^
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -185,7 +185,7 @@ Fixes
   subclasses for mapped traits, used by ``Map``, ``Expression``, ``PrefixMap``.
   (#1091)
 * Fix setting default values via dynamic default methods or overriding trait in
-  subclasses for ``Expression`` and ``AdaptsTo``. (#1088, #1119)
+  subclasses for ``Expression`` and ``AdaptsTo``. (#1088, #1119, #1152)
 
 Deprecations
 ~~~~~~~~~~~~

--- a/traits/ctraits.c
+++ b/traits/ctraits.c
@@ -1833,6 +1833,7 @@ default_value_for(trait_object *trait, has_traits_object *obj, PyObject *name)
                         Py_DECREF(result);
                         return NULL;
                     }
+                    Py_DECREF(value);
                     return result;
                 }
                 else {

